initrd/init

```diff
@@ -337,19 +294,62 @@ for c in `cat /proc/cmdline`; do
 done
 
 cd /
-while :; do
-	for device in ${ROOT:-/dev/[hmnsvx][dmrv][0-9a-z]*}; do
-		check_root $device && break 2
+for device in /dev/[hmnsvx][dmrv][0-9a-z]*; do
+	ln $device /dev/block
+done
 
-		# Also umount /android and /sfs if already mounted
-		mountpoint -q /android && umount /android
-		mountpoint -q /sfs && umount /sfs
-		mountpoint -q /mnt && umount /mnt
-	done
-	sleep 1
-	echo -n .
+# if the android is partition table-based
+IS_ALL_BLOCKDEV=1
+for dev in $SYSTEM $DATA $RECOVERY; do
+	case "$dev" in
+	/dev/*) ;;
+	*) IS_ALL_BLOCKDEV= && break ;;
+	esac
 done
 
+if [ -z "$IS_ALL_BLOCKDEV" ]; then
+
+	ROOT=$(find_device "$ROOT")
+	if [ -z "$ROOT" ]; then
+		source /hooks/0-find-rootdev
+		find_rootdev 10 || debug_shell debug-notfound # maximum 10 iterations
+	else
+		try_mount ro $dev /mnt || debug_shell debug-mountrootfail
+	fi
+
+	if [ -n "$INSTALL" ]; then
+		if [ ! -d "/mnt/isolinux" ] || [ ! -f "/mnt/install.img" ]; then
+			debug_shell debug-missinginstall
+		fi
+	fi
+
+	if [ -n "$iso" -a -e /mnt/$iso ]; then
+		mount --move /mnt /iso
+		mkdir /mnt/iso
+		mount -o loop /iso/$iso /mnt/iso
+	fi
+
+	if [ -z "$SYSTEM" ]; then
+		source /hooks/1-find-system
+		find_system || debug_shell debug-mountsystemfail
```

Comment:
Also this

Reply:
Same almost. If system.?fs, system_$SLOT.img or system is not found, then drop a shell for the user to solve (both in most cases wont likely to happen)